matlab/src/cpp/arrow/matlab/tabular/proxy/schema.cc

```diff
@@ -0,0 +1,182 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/matlab/error/error.h"
+#include "arrow/matlab/tabular/proxy/schema.h"
+#include "arrow/matlab/type/proxy/field.h"
+
+#include "libmexclass/proxy/ProxyManager.h"
+#include "libmexclass/error/Error.h"
+
+#include "arrow/util/utf8.h"
+
+#include <sstream>
+
+namespace arrow::matlab::tabular::proxy {
+
+    Schema::Schema(std::shared_ptr<arrow::Schema> schema) : schema{std::move(schema)} {
+        REGISTER_METHOD(Schema, getFieldByIndex);
+        REGISTER_METHOD(Schema, getFieldByName);
+        REGISTER_METHOD(Schema, getNumFields);
+        REGISTER_METHOD(Schema, getFieldNames);
+        REGISTER_METHOD(Schema, toString);
+    }
+
+    libmexclass::proxy::MakeResult Schema::make(const libmexclass::proxy::FunctionArguments& constructor_arguments) {
+        namespace mda = ::matlab::data;
+        using SchemaProxy = arrow::matlab::tabular::proxy::Schema;
+
+        mda::StructArray args = constructor_arguments[0];
+        const mda::TypedArray<uint64_t> field_proxy_ids_mda = args[0]["FieldProxyIDs"];
+
+        std::vector<std::shared_ptr<arrow::Field>> fields;
+        for (const auto proxy_id : field_proxy_ids_mda) {
+            using namespace libmexclass::proxy;
+            auto proxy = std::static_pointer_cast<arrow::matlab::type::proxy::Field>(ProxyManager::getProxy(proxy_id));
+            auto field = proxy->unwrap();
+            fields.push_back(field);
+        }
+        auto schema = arrow::schema(fields);
+        return std::make_shared<SchemaProxy>(std::move(schema));
+    }
+
+    std::shared_ptr<arrow::Schema> Schema::unwrap() {
+        return schema;
+    }
+
+    void Schema::getFieldByIndex(libmexclass::proxy::method::Context& context) {
+        namespace mda = ::matlab::data;
+        using namespace libmexclass::proxy;
+        using FieldProxy = arrow::matlab::type::proxy::Field;
+        mda::ArrayFactory factory;
+
+        mda::StructArray args = context.inputs[0];
+        const mda::TypedArray<int32_t> index_mda = args[0]["Index"];
+        const auto matlab_index = int32_t(index_mda[0]);
+        // Note: MATLAB uses 1-based indexing, so subtract 1.
+        // arrow::Schema::field does not do any bounds checking.
+        const int32_t index = matlab_index - 1;
+        const auto num_fields = schema->num_fields();
+
+        if (num_fields == 0) {
+            using namespace libmexclass::error;
+            const std::string& error_message_id = std::string{error::ARROW_TABULAR_SCHEMA_NUMERIC_FIELD_INDEX_WITH_EMPTY_SCHEMA};
```

Comment:
Could consider extracting this out into a sub-function. Would make the `make` method smaller.

Reply:
Fixed.